flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,334 @@
+    // task reads from a field, then any known write to the same field has to
+    // finish first), and "write after read" (if this task writes to a field,
+    // then all known reads from the field have to finish before the write).
+    if(privilege_write(P)) {
+      // request future from promise only if required (once for all arguments)
+      if(!future.valid()) {
+        future = promise.get_future();
+      }
+
+      // Any possibly valid field-future must be added as a dependency for this
+      // task (if it has not been added already, is valid, and has not been
+      // marked as ready yet).
+      if(!added_as_dependency && field.future.valid() &&
+         !field.future.is_ready()) {
+        dependencies.push_back(std::move(field.future));
+      }
+
+      // If the task writes to the current argument then we must associate the
+      // future that represents the end of the task execution with the current
+      // argument.
+      field.future = future;
+      field.dep = flecsi::data::dependency::write;
+    }
+    else if(privilege_read(P)) {
+      // request future from promise only if required (once for all arguments)
+      if(!future.valid()) {
+        future = promise.get_future();
+      }
+
+      // If the task reads from the current argument then we must associate the
+      // future that represents the end of the task execution with the current
+      // argument. We have to make sure that possibly more than one read
+      // operation may have to finish before other operations are allowed to go
+      // ahead.
+      if(!field.future.valid() || field.future.is_ready()) {
+        // This is either the first operation using the given field or any
+        // previous operations have already finished.
+        field.future = future;
+      }
+      else if(field.dep != flecsi::data::dependency::write) {
+        // This read operation needs to be added to the list of (read-)
+        // dependencies already existing for the given field.
+        field.future = hpx::when_all(std::move(field.future), future).share();
+      }
+      field.dep = flecsi::data::dependency::read;
+    }
+  }
+
+  // visit for reduction operation
+  template<typename R,
+    typename T,
+    typename Topo,
+    typename Topo::index_space Space>
+  void visit(data::reduction_accessor<R, T> &,
+    const data::field_reference<T, data::dense, Topo, Space> & ref) {
+    static_assert(std::is_same_v<typename Topo::base, topo::global_base>);
+    const field_id_t f = ref.fid();
+    auto & r = *ref.topology();
+
+    // store associated region for bind_accessors
+    regions_partitions.push_back(&r);
+
+    // Reduction input must be treated as a "read/write after write" dependency
+    auto & field = r[f];
+
+    if(field.future.valid()) {
+      flog_assert(field.dep == data::dependency::write,
+        "field reference must represent a 'read after write' dependency");
```

Comment:
Why can't we have a read followed immediately by a reduction?  A reduction just acts like any other read-write.

Reply:
Do we have a test for that?

Reply:
I think `task.cc` is the only test of reduction accessors at all; it runs the reduction twice, but that's a read/write-after-write.

Reply:
Did we decide whether the read-reduce case works with the HPX backend?